### PR TITLE
Issues/#113 Convert txn hash to SHA-3

### PR DESCRIPTION
**Pull request by macterra**
Ready for review

This PR is almost entirely fallout from changing just 3 lines in transaction.pp which change the hash functions from SHA256 to SHA-3.

Because the txn hash changed, we have to generate a new genesis block in chainparams.cpp  and mine a new test blockchain in miner_tests.cpp. In order to do that with CPU mining we (temporarily) lowered the difficulty in chainparams (decreasing the number of leading 0s from 7 to 2).

Lowering the difficulty had an unexpected side-effect of breaking the Dark Gravity Wave unit tests, turns out the difficulty adjustment algorithm had a latent bug where it could not handle low difficulties (high targets) so that was fixed, and while we were fixing that we fixed another bug where the moving average was not being properly calculated (it was slightly wrong because it gave double weight to the most recent block difficulty, and also inefficient in the way it calculate the moving average).

In an earlier PR (#108) a bunch of tests in bloom_tests were disabled because of changes to the merkle tree hash. This PR broke a bunch more in the same file so disabled list was expanded.

### CODE REVIEW

File: src/chainparams.cpp [hunk 1]
```diff
@@ -124,10 +148,31 @@ class CMainParams : public CChainParams {
...
+            genesis = CreateEquibitGenesisBlock(1231006505, nonce, nPoWTarget, 1, 50 * COIN);
+            consensus.hashGenesisBlock = genesis.GetHash();
+            //std::cout << "genesis " << nonce << std::endl;
+
+            if (CheckProofOfWork(consensus.hashGenesisBlock, nPoWTarget, consensus)) {
+                break;
+            }
+        }
+
+        //assert(consensus.hashGenesisBlock == uint256S("0x000000000019d6689c085ae165831e934ff763ae46a2a6c172b3f1b60a8ce26f"));
```
Comment (macterra) [inline src/chainparams.cpp:174]: We'll have to regenerate these asserts eventually fyi @r-soltani

**Comment by macterra** [comment]
TBD: add Equibit copyright to files touched

**Comment by macterra** [comment]
TBD: merge dev, resolve conflicts

**Comment by macterra** [comment]
> 
> 
> TBD: add Equibit copyright to files touched

Done

**Comment by macterra** [comment]
> 
> 
> TBD: merge dev, resolve conflicts

Done

**Review by ealefad — APPROVED** [review]

File: src/test/pow_tests.cpp [hunk 2]
```diff
@@ -209,25 +208,28 @@ BOOST_AUTO_TEST_CASE(GetNextWorkRequired_steady_state)
...
     }
 }
 
 // Simulate a blockchain to check that the difficulty changes in expected ways with the block times
 BOOST_AUTO_TEST_CASE(GetNextWorkRequired_changing)
 {
     const auto chainParams = CreateChainParams(CBaseChainParams::MAIN);
-    Consensus::Params params = chainParams->GetConsensus();
-    const unsigned int nInterval = params.DifficultyAdjustmentInterval() + 1;
+    const Consensus::Params& params = chainParams->GetConsensus();
```
Comment (r-soltani) [inline src/test/pow_tests.cpp:223]: Is the change in the function signature an improvement from Bitcoin's code?

Comment (macterra) [reply src/test/pow_tests.cpp:223]: No this is a new test. The unused variable nInterval was removed. I don't think any signatures have changed in this part of the code relative to the base branch.

**Review by r-soltani — APPROVED** [review]

**Pull request merged**